File: apps/android/app/src/main/java/com/retailmanagement/ui/ManagerScheduleViewModel.kt
package com.retailmanagement.ui

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.retailmanagement.data.api.RetrofitClient
import com.retailmanagement.data.model.*
import kotlinx.coroutines.async
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import java.text.SimpleDateFormat
import java.util.*

class ManagerScheduleViewModel : ViewModel() {

    private val api get() = RetrofitClient.api

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    private val _isActionLoading = MutableStateFlow(false)
    val isActionLoading: StateFlow<Boolean> = _isActionLoading.asStateFlow()

    private val _error = MutableStateFlow<String?>(null)
    val error: StateFlow<String?> = _error.asStateFlow()

    private val _schedule = MutableStateFlow<ScheduleRead?>(null)
    val schedule: StateFlow<ScheduleRead?> = _schedule.asStateFlow()

    private val _shifts = MutableStateFlow<List<ShiftRead>>(emptyList())
    val shifts: StateFlow<List<ShiftRead>> = _shifts.asStateFlow()

    private val _employees = MutableStateFlow<List<StoreEmployeeRead>>(emptyList())
    val employees: StateFlow<List<StoreEmployeeRead>> = _employees.asStateFlow()

    private val _weekStart = MutableStateFlow(currentMonday())
    val weekStart: StateFlow<Date> = _weekStart.asStateFlow()

    private val isoFormat = SimpleDateFormat("yyyy-MM-dd", Locale.getDefault())

    fun currentMonday(): Date {
        val cal = Calendar.getInstance()
        cal.firstDayOfWeek = Calendar.MONDAY
        cal.set(Calendar.DAY_OF_WEEK, Calendar.MONDAY)
        cal.set(Calendar.HOUR_OF_DAY, 0); cal.set(Calendar.MINUTE, 0)
        cal.set(Calendar.SECOND, 0); cal.set(Calendar.MILLISECOND, 0)
        return cal.time
    }

    fun goToPreviousWeek() { _weekStart.value = Date(_weekStart.value.time - 7 * 86400_000L) }
    fun goToNextWeek() { _weekStart.value = Date(_weekStart.value.time + 7 * 86400_000L) }
    fun goToCurrentWeek() { _weekStart.value = currentMonday() }

    fun dayDates(): List<Date> = (0..6).map { Date(_weekStart.value.time + it * 86400_000L) }
    fun dateStr(d: Date): String = isoFormat.format(d)

    fun weekLabel(): String {
        val fmt = SimpleDateFormat("MMM d", Locale.getDefault())
        val end = Date(_weekStart.value.time + 6 * 86400_000L)
        return "${fmt.format(_weekStart.value)} – ${fmt.format(end)}"
    }

    fun loadData(storeId: String) {
        viewModelScope.launch {
            _isLoading.value = true
            _error.value = null
            try {
                val empDef = async { api.getStoreEmployees(storeId) }
                val schedDef = async { api.listSchedules(storeId, isoFormat.format(_weekStart.value)) }

                val empRes = empDef.await()
                val schedRes = schedDef.await()
                _employees.value = empRes.data

                val firstSched = schedRes.data.firstOrNull()
                if (firstSched != null) {
                    val detail = api.getSchedule(storeId, firstSched.id)
                    _schedule.value = detail.data.schedule
                    _shifts.value = detail.data.schedule.shifts
                } else {
                    _schedule.value = null
                    _shifts.value = emptyList()
                }
            } catch (e: Exception) {
                _error.value = e.message ?: "Unknown error"
            }
            _isLoading.value = false
        }
    }

    fun initializeSchedule(storeId: String) {
        viewModelScope.launch {
            _isActionLoading.value = true
            try {
                api.createSchedule(storeId, ScheduleCreate(storeId, isoFormat.format(_weekStart.value)))
                loadData(storeId)
            } catch (e: Exception) {
                _error.value = e.message
            }
            _isActionLoading.value = false
        }
    }

    fun togglePublishStatus(storeId: String) {
        val sched = _schedule.value ?: return
        viewModelScope.launch {
            _isActionLoading.value = true
            try {
                val newStatus = if (sched.status == "draft") "published" else "draft"
                api.updateSchedule(storeId, sched.id, ScheduleUpdate(newStatus))
                loadData(storeId)
            } catch (e: Exception) {
                _error.value = e.message
            }
            _isActionLoading.value = false
        }
    }

    fun saveShift(storeId: String, shiftId: String?, userId: String, date: String, startTime: String, endTime: String, breakMinutes: Int, notes: String?) {
        val sched = _schedule.value ?: return
        viewModelScope.launch {
            _isActionLoading.value = true
            try {
                if (shiftId != null) {
                    api.updateShift(storeId, sched.id, shiftId, ShiftUpdate(startTime, endTime, breakMinutes, notes))
                } else {
                    api.createShift(storeId, sched.id, ShiftCreate(userId, date, startTime, endTime, breakMinutes, notes))
                }
                loadData(storeId)
            } catch (e: Exception) {
                _error.value = e.message
            }
            _isActionLoading.value = false
        }
    }

    fun deleteShift(storeId: String, shiftId: String) {
        val sched = _schedule.value ?: return
        viewModelScope.launch {
            _isActionLoading.value = true
            try {
                api.deleteShift(storeId, sched.id, shiftId)
                loadData(storeId)
            } catch (e: Exception) {
                _error.value = e.message
            }
            _isActionLoading.value = false
        }
    }
}

File: apps/ios/retailmanagement/ViewModels/ManagerScheduleViewModel.swift
//
//  ManagerScheduleViewModel.swift
//  retailmanagement
//

import Foundation
import Observation

@MainActor
@Observable
final class ManagerScheduleViewModel {
    var schedule: ScheduleRead?
    var shifts: [Shift] = []
    var employees: [Employee] = []
    var isLoading = false
    var isActionLoading = false
    var errorMessage: String?

    var weekStart: Date = Calendar.current.dateInterval(of: .weekOfYear, for: Date())?.start ?? Date()

    var weekLabel: String {
        let formatter = DateFormatter()
        formatter.dateFormat = "d MMM"
        let end = Calendar.current.date(byAdding: .day, value: 6, to: weekStart) ?? weekStart
        let yearFmt = DateFormatter()
        yearFmt.dateFormat = "d MMM yyyy"
        return "\(formatter.string(from: weekStart)) – \(yearFmt.string(from: end))"
    }

    var dayDates: [Date] {
        (0..<7).compactMap { Calendar.current.date(byAdding: .day, value: $0, to: weekStart) }
    }

    private static let isoDateFormatter: DateFormatter = {
        let f = DateFormatter()
        f.dateFormat = "yyyy-MM-dd"
        return f
    }()

    func dateString(_ date: Date) -> String {
        Self.isoDateFormatter.string(from: date)
    }

    func shifts(for date: Date) -> [Shift] {
        let key = dateString(date)
        return shifts.filter { $0.shiftDate == key }
    }

    func employee(for userId: String) -> Employee? {
        employees.first { $0.id == userId }
    }

    // MARK: - Week navigation

    func goToPreviousWeek() {
        weekStart = Calendar.current.date(byAdding: .weekOfYear, value: -1, to: weekStart) ?? weekStart
    }

    func goToNextWeek() {
        weekStart = Calendar.current.date(byAdding: .weekOfYear, value: 1, to: weekStart) ?? weekStart
    }

    func goToCurrentWeek() {
        weekStart = Calendar.current.dateInterval(of: .weekOfYear, for: Date())?.start ?? Date()
    }

    // MARK: - Loading

    func loadData(storeId: String) async {
        isLoading = true
        errorMessage = nil
        do {
            async let employeesTask: PaginatedResponse<Employee> = NetworkService.shared.get(
                endpoint: "/api/users/stores/\(storeId)/employees"
            )
            async let schedulesTask: PaginatedResponse<ScheduleRead> = NetworkService.shared.get(
                endpoint: "/api/stores/\(storeId)/schedules",
                queryItems: [URLQueryItem(name: "week_start", value: dateString(weekStart))]
            )
            let employeesResponse = try await employeesTask
            let schedulesResponse = try await schedulesTask
            employees = employeesResponse.data

            if let first = schedulesResponse.data.first {
                let detail: DataResponse<WeeklyScheduleResponse> = try await NetworkService.shared.get(
                    endpoint: "/api/stores/\(storeId)/schedules/\(first.id)"
                )
                schedule = detail.data.schedule
                shifts = detail.data.schedule.shifts
            } else {
                schedule = nil
                shifts = []
            }
        } catch {
            errorMessage = error.localizedDescription
        }
        isLoading = false
    }

    // MARK: - Mutations

    func initializeSchedule(storeId: String) async {
        isActionLoading = true
        do {
            let body = ScheduleCreate(storeId: storeId, weekStart: dateString(weekStart))
            let _: DataResponse<ScheduleRead> = try await NetworkService.shared.post(
                endpoint: "/api/stores/\(storeId)/schedules", body: body
            )
            await loadData(storeId: storeId)
        } catch {
            errorMessage = error.localizedDescription
        }
        isActionLoading = false
    }

    func togglePublishStatus(storeId: String) async {
        guard let sched = schedule else { return }
        isActionLoading = true
        do {
            let newStatus = sched.status == "draft" ? "published" : "draft"
            let _: DataResponse<ScheduleRead> = try await NetworkService.shared.patch(
                endpoint: "/api/stores/\(storeId)/schedules/\(sched.id)",
                body: ScheduleUpdate(status: newStatus)
            )
            await loadData(storeId: storeId)
        } catch {
            errorMessage = error.localizedDescription
        }
        isActionLoading = false
    }

    func saveShift(
        storeId: String, shiftId: String?, userId: String, date: String,
        startTime: String, endTime: String, breakMinutes: Int, notes: String?
    ) async {
        guard let sched = schedule else { return }
        isActionLoading = true
        do {
            if let shiftId {
                let body = ShiftUpdate(startTime: startTime, endTime: endTime, breakMinutes: breakMinutes, notes: notes)
                let _: DataResponse<Shift> = try await NetworkService.shared.patch(
                    endpoint: "/api/stores/\(storeId)/schedules/\(sched.id)/shifts/\(shiftId)", body: body
                )
            } else {
                let body = ShiftCreate(userId: userId, shiftDate: date, startTime: startTime, endTime: endTime, breakMinutes: breakMinutes, notes: notes)
                let _: DataResponse<Shift> = try await NetworkService.shared.post(
                    endpoint: "/api/stores/\(storeId)/schedules/\(sched.id)/shifts", body: body
                )
            }
            await loadData(storeId: storeId)
        } catch {
            errorMessage = error.localizedDescription
        }
        isActionLoading = false
    }

    func deleteShift(storeId: String, shiftId: String) async {
        guard let sched = schedule else { return }
        isActionLoading = true
        do {
            try await NetworkService.shared.deleteNoContent(
                endpoint: "/api/stores/\(storeId)/schedules/\(sched.id)/shifts/\(shiftId)"
            )
            await loadData(storeId: storeId)
        } catch {
            errorMessage = error.localizedDescription
        }
        isActionLoading = false
    }
}
